cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+
+```lisp
+    (decl GetNthArg (u32 Arg) Inst)
+```
+
+that we wish to use as in the example given above, and then we can define
+an external extractor
+
+```lisp
+    (extern extractor GetNthArg get_nth_arg (in out))
+```
+
+which indicates that `get_nth_arg()` will take parameters of `&Inst`
+(the value being extracted) *and* `u32` (the in-arg), and return
+`Option<(Arg,)>`, i.e., *just* the out-args.
+
+In order to use this extractor, to avoid parse ambiguity (i.e., to
+avoid the need for the parser to know argument polarity while it is
+still parsing), we require special syntax for the in-argument so that
+it is parsed as an expression: it must be prepended by `<`. So one
+might use `GetNthArg` as follows:
+
+```lisp
+    (rule (Lower (and
+                   (Inst ...)
+                   (GetNthArg <2 second_arg)))
+          ...)
+```
+
+This functionality is meant to improve the expressive power of the
+DSL, but is not intended to be commonly used outside of "binding" or
+"library" ISLE code. In other words, because it is somewhat
+non-intuitive, it is best to wrap it within friendlier internal
+extractors.
+
+### Mapping Type Declarations to Rust
+
+When we declare a type like
+
+```lisp
+    (decl MyEnum (enum
+                   (A (x u32) (y u32))
+                   (B)))
+```
+
+ISLE will generate the Rust type definition
+
+```rust
+#[derive(Clone, Debug)]
+pub enum MyEnum {
+    A { x: u32, y: u32, },
+    B,
+}
+```
+
+Note that enum variants with no fields take on the brace-less form,
+while those with fields use the named-struct-field `A { x: ... }`
+form. If all variants are field-less, then the type will additionally
+derive `Copy`, `PartialEq`, and `Eq`.
+
+If the type is declared as extern (`(decl MyEnum extern (enum ...))`)
+then the same definition is assumed to exist. Primitives (`(decl u32
+(primitive u32))`) are assumed to be defined already, and are required
+to be `Copy`.
+
+All imported/extern types are pulled in via `use super::*` at the top
+of the generated code; thus, these types should exist in (or be
+re-exported from) the parent module.
+
+### Symbolic Constants
+
+ISLE allows the user to refer to external constants as follows:
+
+```lisp
+    (extern const $I32 Type)
+```
+
+This allows code to refer to `$I32` whenever a value of type `Type` is
+needed, in either a pattern (LHS) or an expression (RHS). These
+constants are pulled in via the same `use super::*` that imports all
+external types.
+  
+### Exported Interface: Functions and Context Trait
+
+The generated ISLE code provides an interface that is designed to be
+agnostic to the embedding application. This means that ISLE knows
+nothing about, e.g., Cranelift or compiler concepts in
+general. Rather, the generated code provides function entry points
+with well-defined signatures based on the terms, and imports the
+extern constructors and extractors via a trait that the embedder must
+implement.
```

Comment:
Maybe name this as the "context trait" here to make the rest of this section more clear?

Reply:
Done.